java-checks/src/main/java/org/sonar/java/checks/LeastSpecificTypeCheck.java

```diff
@@ -0,0 +1,224 @@
+import org.sonar.java.resolve.JavaType;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+
+@Rule(key = "S3242")
+public class LeastSpecificTypeCheck extends IssuableSubscriptionVisitor {
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.METHOD);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+
+    MethodTree methodTree = (MethodTree) tree;
+    Symbol.MethodSymbol methodSymbol = methodTree.symbol();
+    if (!methodSymbol.isPublic()
+      || BooleanUtils.isNotFalse(((MethodTreeImpl) methodTree).isOverriding())
+      || isOverloaded(methodSymbol)) {
+      return;
+    }
+
+    methodTree.parameters().stream()
+      .map(VariableTree::symbol)
+      .filter(p -> p.type().isClass() && !p.type().symbol().isEnum() && !p.type().is("java.lang.String"))
+      .forEach(this::handleParameter);
+  }
+
+  private static boolean isOverloaded(Symbol.MethodSymbol methodSymbol) {
+    return ((Symbol.TypeSymbol) methodSymbol.owner()).lookupSymbols(methodSymbol.name()).size() > 1;
+  }
+
+  private void handleParameter(Symbol parameter) {
+    Type leastSpecificType = findLeastSpecificType(parameter);
+    if (!parameter.type().is(leastSpecificType.fullyQualifiedName())
+      && !leastSpecificType.is("java.lang.Object")) {
+      String suggestedType = leastSpecificType.fullyQualifiedName().replace('$', '.');
+      reportIssue(parameter.declaration(), String.format("Use '%s' here; it is a more general type than '%s'.", suggestedType, parameter.type().name()));
+    }
+  }
+
+  private static Type findLeastSpecificType(Symbol parameter) {
+    List<IdentifierTree> usages = parameter.usages();
+    InheritanceGraph inheritanceGraph = new InheritanceGraph(parameter.type());
+    for (IdentifierTree usage : usages) {
+      Tree parent = usage.parent();
+      while (parent != null && !parent.is(Tree.Kind.ARGUMENTS, Tree.Kind.METHOD_INVOCATION, Tree.Kind.EXPRESSION_STATEMENT, Tree.Kind.FOR_EACH_STATEMENT)) {
+        parent = parent.parent();
+      }
+      if (parent == null || parent.is(Tree.Kind.EXPRESSION_STATEMENT, Tree.Kind.ARGUMENTS)) {
+        return parameter.type();
+      }
+      if (parent.is(Tree.Kind.FOR_EACH_STATEMENT)) {
+        // find Iterable.iterator() method
```

Comment:
Please extract this into a dedicated well named method rather than adding a comment. 
What if the parameter type iterated upon is an array ?

Reply:
Array parameters are excluded before (line #70)